keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+Extending image pulls to specify an optional runtime class would greatly help to extend windows platform matcher behavior depending on whether hyperV or process isolation is being used. Additionally, having support to pull images per runtime class might also help projects like supporting different snapshotters with different runtimes as well.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+This KEP describes the detailed design and major API changes needed in kubelet, CRI and containerd to support image pull with optional runtime class option.
+Goal is to keep the functionality same as it is today if user does not use the runtime class option with image pull.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### CRI/kubelet changes:
+
+With this feature, multiple versions of the same image can exist on a single node. For exmaple, python image can exist on the node for WS2022 (20348 build number) and WS2019(17703 build number) depending on the guestOSVersion mentioned in containerd's toml file for a runtime. Therefore, it is important for kubelet, containerd and other components to identify this. The following sections details the major changes that would need to be made to the different components to support this.
+
+#### ImageSpec
+Include field to specify optional runtime class to use to pull image
+
+
+**Proposed:**
+
+<pre>
+// ImageSpec is an internal representation of an image.
+message ImageSpec {
+    // Container's Image field (e.g. imageID or imageDigest).
+    string image = 1;
+    // Unstructured key-value map holding arbitrary metadata.
+    // ImageSpec Annotations can be used to help the runtime target specific
+    // images in multi-arch images.
+    map<string, string> annotations = 2;
+    // The container image reference specified by the user (e.g. image[:tag] or digest).
+    // Only set if available within the RPC context.
+    string user_specified_image = 18;
+    // Named runtime configuration to use for pulling the image.
+    // If the runtime handler is unknown, this request should be rejected.  An
+    // empty string should select the default handler, equivalent to the
+    // behavior before this feature was added.
+    <mark>string runtime_handler = 19;</mark>
```

Comment:
Is this the only change required to also address the `snapshotter per runtimeclass` scenario?

Reply:
No, CRI has another store for storing snapshots related data. I haven't looked into it in detail but this is one of the main files that would need changes - https://github.com/containerd/containerd/blob/main/pkg/cri/store/snapshot/snapshot.go .

This KEP does not intend to cover snapshots per runtime class changes. It is a project I have heard folks are interested in several meetings so I just added it as one of the projects that would be benefitted from this image pull per runtime class. I will call it out in the KEP specifically to avoid any confusion.

Reply:
Can we say that snapshots per runtime class changes are out of scope for this KEP and say something like implementation details will be handled by each container runtime?